LRUStoreCache

ruaridhg opened the pull request:
[Description of PR]

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in `docs/user-guide/*.rst`
* [ ] Changes documented as a new file in `changes/`
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,705 @@
+            return f.write(view)
+
+
+class LRUStoreCache(Store):
```
Could you add a note about what is cached in the store; is it the raw bytes that come from the store, or is it the array data obtained after de-compressing those bytes?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,705 @@
+            return f.write(view)
+
+
+class LRUStoreCache(Store):
```
The docstring should mention whether it's possible to write data with this store, or if it's a read only store. (in my head it makes sense for this to just be a read only store, but maybe I'm wrong?)

ruaridhg replied:
There's a read-only option, but it's also got a write option and I've updated the docstring to describe this

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,705 @@
+    ----------
+    store : Store
+        The store containing the actual data to be cached.
+    max_size : int
+        The maximum size that the cache may grow to, in number of bytes. Provide `None`
+        if you would like the cache to have unlimited size.
```
I would be tempted to remove `None` as an option, because having an unlimited cache size seems like it has potential to go wrong for users without realising by growing beyond the size of available RAM. What do you think?

ruaridhg replied:
I agree, I've removed `None`

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,705 @@
+
+    root: Path
+
+    def __init__(self, store: Store, max_size: int | None, **kwargs: Any) -> None:
```
`max_size` should definitely be keyword-only here to prevent calls like `LRUCacheStore(store, 103094)`, where it wouldn't be clear what the number means.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,705 @@
+    ----------
+    store : Store
+        The store containing the actual data to be cached.
+    max_size : int
```
What happens when `max_size` is a very small number (say, 1 byte)? Does the cache still work somehow? Is there a practical minimum value of `max_size` that needs to be set for the caching to be effective?

ruaridhg replied:
Added some suggested values to the docstring

dstansby commented on tests/test_store/test_cache.py:
When I run the pre-commit checks, there are lots of `mypy` errors in this file that need fixing.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,705 @@
+    --------
+    The example below wraps an S3 store with an LRU cache::
+
+        >>> import s3fs
```
Would it be possible to change this example to work with a `LocalStore` instead? Most folks won't have access to an S3 bucket and I figure the examples should be code most people can run.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,705 @@
+    ----------
+    store : Store
+        The store containing the actual data to be cached.
+    max_size : int
```
I think typically users will think of cache sizes in MB (or even GB), so it's a bit of a pain having to specify this in bytes. I don't have a good suggestion, and it could definitely be beyond scope of this PR, but if you can think of an easy way to accept something like `max_size="16MB"` that would really improve usability.

ruaridhg replied:
Will think about this is a future commit

dstansby commented on changes/3357.feature.rst:
```diff
@@ -0,0 +1 @@
+Add LRUStoreCache to Zarr 3.0
```
Don't need to specify "3.0" here, the changelog entries appear in a section that has the version: https://zarr.readthedocs.io/en/stable/release-notes.html.

dstansby replied:
It would be nice to add a short explanation of what "LRUStoreCache" is, and what it's useful for, so users reading the release notes can get exited about/understand what's new.

dstansby commented on docs/user-guide/lrustorecache.rst:
```diff
@@ -0,0 +1,218 @@
+   >>>
+   >>> # Create a local store and wrap it with LRU cache
+   >>> local_store = zarr.storage.LocalStore('test.zarr')
+   >>> cache = zarr.storage.LRUStoreCache(local_store, max_size=2**28)  # 256MB cache
```
```suggestion
   >>> cache = zarr.storage.LRUStoreCache(local_store, max_size=1024 * 1024 * 256)  # 256MB cache
```
Expanding out the maths makes it more obvious that it's 256 MB.

dstansby commented on docs/user-guide/lrustorecache.rst:
```diff
@@ -0,0 +1,218 @@
+   >>> elapsed_nocache = time.time() - start
+   >>>
+   >>> speedup = elapsed_nocache/elapsed_cache
+   >>> # Speedup typically ranges from 1.5x to 5x depending on system
```
Where do these numbers come from? Have you tested on multiple systems?

ruaridhg replied:
Ah this was when I was trying it locally multiple times, but then the output expected nothing so I removed the print statement in this example

dstansby commented on docs/user-guide/lrustorecache.rst:
```diff
@@ -0,0 +1,218 @@
+   >>> # Typical hit ratio is > 50% with repeated access patterns
+
+Cache Management
+---------------
```
These title underlines need to match the length of the title, otherwise the documetation build will be sad 😢

dstansby commented on docs/user-guide/lrustorecache.rst:
```diff
@@ -0,0 +1,218 @@
+3. **Monitor cache statistics**: Use hit/miss ratios to tune cache size and access patterns
+4. **Consider data locality**: Group related data accesses together to improve cache efficiency
+
+Working with Different Store Types
```
I don't really see the need to have this seciton - it's clear from higher up the guide that different stores can be used, and there's even expamples of using two different stores above here.

dstansby commented on docs/user-guide/lrustorecache.rst:
```diff
@@ -0,0 +1,218 @@
+1. **Size the cache appropriately**: Set ``max_size`` based on available memory and expected data access patterns
+2. **Use with remote stores**: The cache provides the most benefit when wrapping slow remote stores
+3. **Monitor cache statistics**: Use hit/miss ratios to tune cache size and access patterns
+4. **Consider data locality**: Group related data accesses together to improve cache efficiency
```
What does "data locality" mean? And what does "Group related data accesses" together mean in practice?

dstansby commented on pyproject.toml:
```diff
@@ -392,6 +392,7 @@ filterwarnings = [
     "ignore:Unclosed client session <aiohttp.client.ClientSession.*:ResourceWarning"
 ]
 markers = [
+    "asyncio: mark test as asyncio test",
```
Did you need to add this to stop some tests failing (or pytest in general failing)? It seems unrelated to the tests you added, so I'm a bit surprised.

ruaridhg replied:
I can remove it for now, but yes I think there were some failing tests within the github actions when I didn't include it. Tests pass locally fine without it.

dstansby commented on src/zarr/storage/__init__.py:
```diff
@@ -4,6 +4,7 @@
 from typing import Any
 
 from zarr.errors import ZarrDeprecationWarning
+from zarr.storage._cache import LRUStoreCache
```
You'll need to add this to `__all__` below so it turns up in the documentation.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+from zarr.core.buffer.core import default_buffer_prototype
+from zarr.storage._utils import normalize_path
+
+ByteRequest: TypeAlias = RangeByteRequest | OffsetByteRequest | SuffixByteRequest
```
It probably makes sense to define this in `zarr.abc.store` and export it from there, since it only uses types defined there anyway.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+ByteRequest: TypeAlias = RangeByteRequest | OffsetByteRequest | SuffixByteRequest
+
+
+def buffer_size(v: Any) -> int:
```
Why does this function accept anything, instead of just accepting a buffer object? Would it be possible to rewrite so it just accepts a buffer? Doing lots of `hasattr` checks seems quite brittle and likely to fail with some edge cases, so if it's possible to narrow the input type here that would make for more robust code.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+    return prefix
+
+
+def _listdir_from_keys(store: Store, path: str | None = None) -> list[str]:
```
Instead of accepting `str | None` here for path, could you accept just `str` - this would allow you to remove the `_path_to_prefix()` function.

dstansby replied:
This needs a short docstring explainig what it does.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+def listdir(store: Store, path: Path | None = None) -> list[str]:
+    """Obtain a directory listing for the given path. If `store` provides a `listdir`
+    method, this will be called, otherwise will fall back to implementation via the
+    `MutableMapping` interface."""
```
```suggestion
    """
    Obtain a directory listing for the given path.
    
    If `store` provides a `listdir`
    method, this will be called, otherwise will fall back to implementation via the
    `MutableMapping` interface.
    """
```
If possible, [the first line of a docstring should be a one line summary.](https://numpydoc.readthedocs.io/en/latest/format.html#short-summary)

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+    method, this will be called, otherwise will fall back to implementation via the
+    `MutableMapping` interface."""
+    path_str = normalize_path(path)
+    if hasattr(store, "listdir"):
```
Instead of using `hasattr` here, you should use the property on `Store` that allows you to check if it is listable.

ruaridhg replied:
Addressed this now.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        result = store.listdir(path_str)
+        return [str(item) for item in result]  # Ensure all items are strings
+    else:
+        # slow version, iterate through all keys
```
You haven't actually impelmented anything in this branch to list the directory?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        return _listdir_from_keys(store, path_str)
+
+
+def _get(path: Path, prototype: BufferPrototype, byte_range: ByteRequest | None) -> Buffer:
```
Needs a short docsting.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        return prototype.buffer.from_bytes(f.read())
+
+
+def _put(
```
Needs a short docstring.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+    """Storage class that implements a least-recently-used (LRU) cache layer over
+    some other store. Intended primarily for use with stores that can be slow to
+    access, e.g., remote stores that require network communication to store and
+    retrieve data.
```
```suggestion
    """
    Storage class that implements a least-recently-used (LRU) cache layer over
    some other store.
    
    Intended primarily for use with stores that can be slow to
    access, e.g., remote stores that require network communication to store and
    retrieve data.
```

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        This parameter is required to prevent unbounded memory growth.
+
+        Recommended minimum values:
+        - For small metadata files: 1MB (1024 * 1024)
```
I don't understand this - doesn't the cache read array data and not metadata files?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+
+        Recommended minimum values:
+        - For small metadata files: 1MB (1024 * 1024)
+        - For chunk caching: 10MB (1024 * 1024 * 10)
+        - For general use: 256MB (1024 * 1024 * 256)
+        - For high-performance applications: 1GB (1024 * 1024 * 1000)
```
I don't really understand these recommended values either - What logic did you use to decide on these?

ruaridhg replied:
I'll remove this and add functionality for more user-friendly max_size inputs

dstansby replied:
I think it might be worthwhile adding that functinality later to keep the size of this PR down. I was just confused here by how you chose the recommendations on cache size.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        - For general use: 256MB (1024 * 1024 * 256)
+        - For high-performance applications: 1GB (1024 * 1024 * 1000)
+
+        Values smaller than your typical chunk size will result in most data
```
If you set a small value, and then try and get data from an array with a larger chunk size, is some sort of warning issued? If not it would be good to add this warning and a test for it.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+
+    root: Path
+
+    def __init__(self, store: Store, *, max_size: int, **kwargs: Any) -> None:
```
The parameter doc above needs to document where the `kwargs` go

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            raise ValueError("max_size must be a positive integer (bytes)")
+
+        # Extract and handle known parameters
+        read_only = kwargs.get("read_only", getattr(store, "read_only", False))
```
Instead of getting `read_only` from kwargs, it makes for more obvious and cleaner code to explicitly have `read_only` as a function argument.

dstansby replied:
I would suggest that this store just directly takes the `read_only` state from the underlying store, instead of allowing the user to specify it. That way we avoid having to think about what to do when the two stores have different read only values.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        underlying_store = self._store.with_read_only(read_only)
+        return LRUStoreCache(underlying_store, max_size=self._max_size, read_only=read_only)
+
+    def _normalize_key(self, key: Any) -> str:
```
```suggestion
    def _normalize_key(self, key: str | Path) -> str:
```
Seems like the correct more tight typing for this?

dstansby replied:
Also, this doesn't need to be a method because it doesn't use `self`, it can just live as a helper funciton somewhere.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            return str(key)
+        return str(key)
+
+    def __getstate__(
```
Can you remind me what the purpose of `__getstate__` and `__setstate__` is?

ruaridhg replied:
I believe it's for pickling/unpickling and it appears in LRUCache for v2

dstansby replied:
> it appears in LRUCache for v2

This in itself isn't a reason for including it in the new implementation 😄

Is pickling/unpickling a property that all stores must have in zarr-python v3, or are there existing tests that other stores can be pickled/unpicked? If not, then for simplicitly it might be that you can remove these methods.

ruaridhg replied:
A pre-existing test fails in `testing/store.py` when I remove them:
```
async def test_serializable_store(self, store: S) -> None:
        new_store: S = pickle.loads(pickle.dumps(store))
        assert new_store == store
        assert new_store.read_only == store.read_only
        # quickly roundtrip data to a key to test that new store works
        data_buf = self.buffer_cls.from_bytes(b"\x01\x02\x03\x04")
        key = "foo"
        await store.set(key, data_buf)
        observed = await store.get(key, prototype=default_buffer_prototype())
        assert_bytes_equal(observed, data_buf)
```

dstansby replied:
👍 in that case I think it's good to keep them.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+    def __iter__(self) -> Iterator[str]:
+        return self.keys()
+
+    def __contains__(self, key: Any) -> bool:
```
```suggestion
    def __contains__(self, key: str | Path) -> bool:
```
I think this should have tighter typing?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        with self._mutex:
+            if key not in self._contains_cache:
+                # Handle both Store objects and dict-like objects
+                if hasattr(self._store, "__contains__"):
```
`__contains__` is not a method in the Store base calss, so I don't think this branch will ever be hit?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+                else:
+                    # Fallback for stores without __contains__
+                    try:
+                        if hasattr(self._store, "__getitem__"):
```
Stores also don't have a `__getitem__` method, so it feels like none of these paths will be hit... Is there a reason you're not using the methods defined on the base Store class here (specifically, I would expect `Store.get()` to be used to get a value from the store)?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        await self._store.clear()
+        self.invalidate()
+
+    def keys(self) -> Iterator[str]:
```
Needs a docstring

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+                        self._contains_cache[key] = False
+            return bool(self._contains_cache[key])
+
+    async def clear(self) -> None:
```
Needs a docstring

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+    def _keys(self) -> list[str]:
+        if self._keys_cache is None:
+            # Handle both Store objects and dict-like objects
+            if hasattr(self._store, "keys") and callable(self._store.keys):
```
`Store` doesn't have a `keys` method or attribute - is this branch ever being hit?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        await self._store.clear()
+        self.invalidate()
+
+    def keys(self) -> Iterator[str]:
```
What's the motivation for having a method that returns all the keys in the underlying store? I can't think of a use case off the top of my head, but I'm sure I could be missing something!

ruaridhg replied:
Again something that was in v2, happy to remove!

dstansby replied:
If there aren't similar methods in other v3 stores, they I'd say remove - if there are for consistency then maybe we could keep?

ruaridhg replied:
They aren't in any other v3 stores so removed.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+                self._keys_cache = []
+        return self._keys_cache
+
+    def listdir(self, path: Path | None = None) -> list[str]:
```
Needs a docstring

dstansby replied:
Why allow `path = None` here? I think it would be clearer to just accept `Path` as the type, and the expect users to pass `Path("/")` if they want to list the root path.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+                self._listdir_cache[path_key] = listing
+                return listing
+
+    async def getsize(self, key: str) -> int:
```
Needs a docstring

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        _, v = self._values_cache.popitem(last=False)
+        return v
+
+    def _accommodate_value(self, value_size: int) -> None:
```
Needs docstring

ruaridhg replied:
Does this need a docstring since it's a private method?

dstansby replied:
I think for developer understanding it's useful to have a short comment or docstring (one sentence) explaining what each method does.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            v = self._pop_value()
+            self._current_size -= buffer_size(v)
+
+    def _cache_value(self, key: str, value: Any) -> None:
```
Needs docstring

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            cache_value = value
+
+        value_size = buffer_size(cache_value)
+        # Check if value exceeds max size - if so, don't cache it
```
Do you think it's worth emitting a warning if a value is being skipped and not cached?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        with self._mutex:
+            self._values_cache.clear()
+
+    def invalidate_keys(self) -> None:
```
Instead of having two methods with near identical names, it might be clearer to just have a `invalidate_keys()` mehod and merge them. At the moment having two makes it a bit confusing which one should be used where.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        self._contains_cache.clear()
+        self._listdir_cache.clear()
+
+    def _invalidate_value(self, key: Any) -> None:
```
Same comment re. two similar methods as above.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            value = self._values_cache.pop(cache_key)
+            self._current_size -= buffer_size(value)
+
+    def __getitem__(self, key: Any) -> Any:
```
Shouldn't `key` always be a string?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+
+        except KeyError:
+            # cache miss, retrieve value from the store
+            if hasattr(self._store, "__getitem__"):
```
`Store` class doesn't have a `__getitem__` method

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        return value
+
+    def __setitem__(self, key: str, value: Buffer) -> None:
+        if hasattr(self._store, "__setitem__"):
```
Store class doesn't have a `__setitem__` method

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            self._cache_value(self._normalize_key(key), value)
+
+    def __delitem__(self, key: Any) -> None:
+        if hasattr(self._store, "__delitem__"):
```
Store class doesn't have a `__delitem__` method

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        self._check_writable()
+
+        # Delegate to the underlying store for actual deletion
+        if hasattr(self._store, "delete"):
```
Instead of hasattr, this should uset the `supports_deletes` property

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            await self._store.delete(key)
+        else:
+            # Fallback for stores that don't have async delete
+            del self._store[key]  # type: ignore[attr-defined]
```
If the store doesn't support deletes, doesn't that mean deletes just don't work and this should error?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+
+    async def exists(self, key: str) -> bool:
+        # Delegate to the underlying store
+        if hasattr(self._store, "exists"):
```
All stores define `exists()` (it's an abstract method on `Store`), so no need for this logic.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+                except KeyError:
+                    return False
+
+    async def _set(self, key: str, value: Buffer, exclusive: bool = False) -> None:
```
Needs docstring

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+        self._check_writable()
+
+        # Delegate to the underlying store
+        if hasattr(self._store, "set"):
```
All stores have a `set()` method

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+            self._invalidate_keys()
+            self._cache_value(self._normalize_key(key), value)
+
+    def __delitem__(self, key: Any) -> None:
```
Why implement these special methods (including `__setitem__`? No other stores seem to implement them, and have other methods that do the same thing instead.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,735 @@
+
+        # For byte_range requests, don't use cache for now (could be optimized later)
+        if byte_range is not None:
+            if hasattr(self._store, "get") and callable(self._store.get):
```
All stores have a `get()` method.

dstansby commented on docs/user-guide/lrustorecache.rst:
```diff
@@ -0,0 +1,186 @@
+``None`` for unlimited cache size (use with caution).
+
+Performance Benefits
+-------------------
```
There are still some titles that don't have matching underline lengths in this file.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+    Extract directory listing from store keys by filtering keys that start with the given path.
+    """
+    children: set[str] = set()
+    # Handle both Store objects and dict-like objects
```
Why are you handling both Store and dict-like objects, when the type of store in the signature only allows Store objects?

ruaridhg replied:
Remove handling of both types of objects.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+                    self._keys_cache = []
+        return self._keys_cache
+
+    def listdir(self, path: Path) -> list[str]:
```
This whole method can be deleted, because it's made redundant by `list_dir()`, which is the method that the `Store` base class says has to be implemented.

ruaridhg replied:
Deleted this.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+    return sorted(children)
+
+
+def listdir(store: Store, path: Path | None = None) -> list[str]:
```
This whole function can be deleted, because it's only used in `listdir()` below, which can also be delted (see other comment below).

ruaridhg replied:
Deleted this.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+    supports_writes: bool = True
+    supports_deletes: bool = True
+    supports_partial_writes: bool = True
+    supports_listing: bool = True
```
I think this will not always be `True`, because it depends on if the underlying store supports listing. In fact, I think *all* these properties will depend on the underlying store. So they should be re-implemented as properties that return the values from the underlying store.

ruaridhg replied:
Made into properties that use underlying store.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+    supports_partial_writes: bool = True
+    supports_listing: bool = True
+
+    root: Path
```
What does this attribute do? It does not seem to be used anywhere apart from being set in `__init__`

ruaridhg replied:
Removed this.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+            raise ValueError("max_size must be a positive integer (bytes)")
+
+        # Always inherit read_only state from the underlying store
+        read_only = getattr(store, "read_only", False)
```
Instead of using `getattr` here, it's safer to just access the attribute directly (ie, do `read_only=store.read_only`). This way the type will be automatically inferred, and an error will be raised if for some reason `store` does not have a `read_only` property (which it should, because it's defined on the `Store` base class.

ruaridhg replied:
Yep changed this now.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        # Always inherit read_only state from the underlying store
+        read_only = getattr(store, "read_only", False)
+
+        # Call parent constructor with read_only parameter
```
This comment doesn't describe anything or add any context beyond the actual line of code, so I'd recommend removing.

ruaridhg replied:
Removed.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+    @classmethod
+    async def open(cls, store: Store, *, max_size: int, **kwargs: Any) -> "LRUStoreCache":
+        """
+        Create and open the LRU cache store.
```
```suggestion
        Create and open a new LRU cache store.
```

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        max_size : int
+            The maximum size that the cache may grow to, in number of bytes.
+        **kwargs : Any
+            Additional keyword arguments passed to the store constructor.
```
Because `max_size` is explicitly  in the signature to `open()`, which is the only keyword argument to `__init__`, `kwargs` don't need to be added to the signature.

ruaridhg replied:
Removed kwargs.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        LRUStoreCache
+            The opened cache store instance.
+        """
+        # Handle read_only parameter if provided
+        if "read_only" in kwargs:
+            read_only = kwargs.pop("read_only")
+            store = store.with_read_only(read_only)
```
I don't think this is the right place to handle read only. If a user wants the LRUStore to be read only, they should pass in a store that is already read-only (and be responsible themselves for setting this).

ruaridhg replied:
It's handled in other stores here, I've modified the implementation so that the cache becomes read_only if the underlying store is read_only. Otherwise pre-existing tests are failing. in `test_store_open_read_only` in `testing/store.py`

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        underlying_store = self._store.with_read_only(read_only)
+        return LRUStoreCache(underlying_store, max_size=self._max_size)
+
+    def _normalize_key(self, key: str | Path) -> str:
```
I think (you should check) in all cases where this is called, `key` is already a `str`, making this method redundant and meaning it can be removed.

ruaridhg replied:
Changed to `cache_key = key` without normalize_key function so that it's obvious that it's a cached key but can change if cleaner.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+    def __iter__(self) -> Iterator[str]:
+        return self.keys()
+
+    def __contains__(self, key: str | Path) -> bool:
```
No other stores define `__contains__`, so unless there's a good reason for implementing it here I'd get rid of this method for simplicity.

ruaridhg replied:
Removed.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        This operation clears both the underlying store and invalidates
+        all cached data to maintain consistency.
+        """
+        # Check if store is writable
+        self._check_writable()
```
I think these lines can be removed, because the appropriate error will be raised by `self._store.clear()` if the store is not writeable.

ruaridhg replied:
Removed.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+    def _keys(self) -> list[str]:
+        if self._keys_cache is None:
+            # Check if it's a Store object vs dict-like object
+            if hasattr(self._store, "supports_listing"):
```
All stores have a `supports_listing` property (you can see this defined on the base `Store` class), so this if check is redundant because it always returns `True`.

ruaridhg replied:
Removed this.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+                    # Store doesn't support listing
+                    self._keys_cache = []
+            else:
+                # It's a dict-like object (for tests)
```
Why are you checking for dict-like objects here (and elsewhere)? `self._store` is always an instance of `Store`, so I don't understand the need to implement this logic.

ruaridhg replied:
Also removed.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        """
+        Get the size in bytes of the value stored at the given key.
+        """
+        return await self._store.getsize(key)
```
Shouldn't you be checking if the value is in the cache, and if it is getting the size from the value in the cache, instead of going into the underlying store every time?

ruaridhg replied:
True, re-implemented this.

dstansby replied:
Noice 👍 A follow up question: if the size of the key is being retrieved from the store, is it worth also caching the value at that point? If `_store.getsize(key)` retrieves the full value and then gets it's length, then I would say it's worth adding the value to the cache, because the cost of getting the value has been paid.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        # remove the first value from the cache, as this will be the least recently
+        # used value
+        _, v = self._values_cache.popitem(last=False)
+        return cast(bytes, v)
```
If you're always expecting bytes in `_values_cache`, then the typing of `_values_cache` should be updated above to be dict[str, bytes] and this cast removed.

ruaridhg replied:
Changed.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+            self._contains_cache.clear()
+            self._listdir_cache.clear()
+
+    def _invalidate_keys_unsafe(self) -> None:
```
Instead of defining this and `invalidate_keys()` and duplicating code, I would keep just `invalidate_keys()`, and when you need to call it just always call it outsdie a `with self._mutex:` block (because it handles the mutex iteself).

ruaridhg replied:
Fair enough, implemented this.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        self._contains_cache.clear()
+        self._listdir_cache.clear()
+
+    def _invalidate_value_unsafe(self, key: Any) -> None:
```
Same comment as above.

ruaridhg replied:
Done.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        self._check_writable()
+
+        # Check if it's a Store object vs dict-like object
+        if hasattr(self._store, "supports_listing"):
```
This is redundant, all stores define the `supports_listing` property.

ruaridhg replied:
Removed.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+        with self._mutex:
+            self._invalidate_keys_unsafe()
+            cache_key = self._normalize_key(key)
+            self._invalidate_value_unsafe(cache_key)
```
Here I think it's easier to just do `_invalidate_keys()` and then `_invalidate_value()` without the mutex block.

ruaridhg replied:
Done.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,698 @@
+
+    async def exists(self, key: str) -> bool:
+        # Delegate to the underlying store
+        return await self._store.exists(key)
```
Shouldn't this be checking the cache first instead of checking the store every time? Otherwise it defeats the point of having the cache!

ruaridhg replied:
True, matching with getsize now.

dstansby replied:
👏

dstansby reviewed (commented):
Some general comments:

- There are several `if hasattr(...)` blocks that are redundant, because all stores define those attributes. You can check which properties/attributes/methods all stores have at https://zarr.readthedocs.io/en/stable/api/zarr/abc/store/index.html#zarr.abc.store.Store. These redudnant blocks should be simplified. I started commenting on these, but there's enough that I stopped, so you'll have to find them all.
- There are several branches that claim to handle "dict-like" objects, but `self._store` is always a store. So these branches can be removed. I started commenting on these, but there's enough that I stopped, so you'll have to find them all.
- There are a few methods that are redundant or duplicate functionality, and should be removed (see inline comments)

ruaridhg commented:
> Some general comments:
> 
> * There are several `if hasattr(...)` blocks that are redundant, because all stores define those attributes. You can check which properties/attributes/methods all stores have at https://zarr.readthedocs.io/en/stable/api/zarr/abc/store/index.html#zarr.abc.store.Store. These redudnant blocks should be simplified. I started commenting on these, but there's enough that I stopped, so you'll have to find them all.
> * There are several branches that claim to handle "dict-like" objects, but `self._store` is always a store. So these branches can be removed. I started commenting on these, but there's enough that I stopped, so you'll have to find them all.
> * There are a few methods that are redundant or duplicate functionality, and should be removed (see inline comments)

I've addressed a lot of the inline comments and removed the dict-like branches, but I'm getting lots of the same mypy errors when removing if hasattr blocks either:
Function "list" could always be true in boolean context  [truthy-function] - when I changed it to if self._store.list:
Statement is unreachable  [unreachable] - when I changed it to if callable(self._store.keys):

```
async def list(self) -> AsyncIterator[str]:
        # Delegate to the underlying store
        if self._store.list:
            async for key in self._store.list():
                yield key
        else:
            # Fallback for stores that don't have async list
            if callable(self._store.keys):
                for key in list(self._store.keys()):
                    yield key
```

Will we always be delegating to the underlying store in these cases? In which instance, I can get rid of the if-else statement which resolves the mypy issues. If we want to have the underlying store or the cache for these methods then I'm not sure how to keep the if-else statement without running into mypy issues.

ruaridhg reviewed (commented):
I've commented and/or implemented changes for the inline comments.

dstansby commented:
> Function "list" could always be true in boolean context [truthy-function] - when I changed it to if self._store.list:
Statement is unreachable [unreachable] - when I changed it to if callable(self._store.keys):

`.list` is a function, so I'm not actually sure how `if self._store.list:` is evaluated. Instead, you can use [`.supports_listing`](https://zarr.readthedocs.io/en/stable/api/zarr/abc/store/index.html#zarr.abc.store.Store.supports_listing) to check if the store supports listing or not, and if it doesn't an error should be raised by `LRUCacheStore` because it won't be possible to list the underlying store.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+        if hasattr(store, "root"):
+            self.root = store.root
+        else:
+            self.root = Path("/")  # Default root path
```
I don't think these lines are used, so can be safely deleted?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+            self.root = Path("/")  # Default root path
+
+    @classmethod
+    async def open(cls, store: Store, *, max_size: int, read_only: bool = False) -> "LRUStoreCache":
```
```suggestion
    async def open(cls, store: Store, *, max_size: int, read_only: bool = False) -> Self:
```
Since this returns a instance of `cls`, the correct typing annotation here is to return `Self` (which is the type of `cls`). This means any sub-classes have the correct typing. (Although I'm not sure why one would subclass...). You'll need to import `Self` from `typing` at the top too.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+        self._store = store
+        self._max_size = max_size
+        self._current_size = 0
+        self._keys_cache: list[str] | None = None
```
It seems like the `_keys_cache` variable isn't used in any of the caching logic, so can be removed?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+        self._mutex = Lock()
+
+    def __len__(self) -> int:
+        return len(self._keys())
```
If you get rid of `_keys_cache` (as above comment), this could be changed to `len(self._values_cache)`, which would then give you the number of values in the cache. I *think* that's the intention of the return value here?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+        await self._store.clear()
+        self.invalidate()
+
+    def _keys(self) -> list[str]:
```
If you get rid of `_keys_cache`, I think this method could go too, since it's only used in `__len__`.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+        _, v = self._values_cache.popitem(last=False)
+        return v
+
+    def _accommodate_value(self, value_size: int) -> None:
```
```suggestion
    def _accommodate_value(self, value_size: int) -> None:
    # Remove items from the cache until there's enough room for a value
```

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+            self._values_cache.clear()
+            self._current_size = 0
+
+    def invalidate_keys(self) -> None:
```
If `_keys_cache` isn't needed, this method can go too.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+        # Not in cache, delegate to underlying store
+        return await self._store.exists(key)
+
+    async def _set(
```
Since this is only used in `set()`, it would be less mis-direction when reading the code to remove this and put the implementation directly in the `set()` method.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+
+        # Update cache
+        self.invalidate_keys()
+        self.invalidate_values()
```
If only a single key is being set, why does the whole cache need to be invalidated? My thinking is one only needs to update the cache value at `key` if it's present, or add it if it's not.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+                # For now, just return None for byte range requests on sync stores
+                return None
+
+        try:
```
A bit safter to do `if cache_key in self._values_cache` here instead of try/except. That way any other KeyErrors raised in the block will be correctly raised, instead of being silently caught.

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,532 @@
+        if byte_range is not None:
+            try:
+                if prototype is None:
+                    prototype = default_buffer_prototype()
```
These lines are repeated underneath in different blocks - could it be de-duplicated by being put at the top outside the if/else block?

dstansby commented on tests/test_store/test_cache.py:
```diff
@@ -0,0 +1,409 @@
+from zarr.testing.store import StoreTests
+
+
+class CountingDict(dict[Any, Any]):
```
Checking how many times the different methods have been called is a really neat idea! This currently needs some fixing though, because you should be testing how many times method on a `Store` obejct is called, but this is not a `Store` object (although it sort of looks like one).

So my recommendation to fix this is to remove `CountingDict`, and instead implement a store that is a thin wrapper of a `MemoryStore`, that also contains logic to track method calls. Something like:


```python
class CounterStore(MemoryStore):
    """
    A thin wrapper of MemoryStore to count different method calls for testing.
    """

    def __init__(self) -> None:
        super().__init__()
        self.counter: Counter[tuple[str, Any] | str] = Counter()

    async def clear(self) -> None:
        self.counter["clear"] += 1
        # docstring inherited
        self._store_dict.clear()

# TODO: implement other methods that should be tracked
```

Does that make sense?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,502 @@
+                    with self._mutex:
+                        if cache_key not in self._values_cache:
+                            self._cache_value(cache_key, value)
+            except Exception:
```
Why are all exceptions being caught here, without being re-raised?

dstansby commented on src/zarr/storage/_cache.py:
```diff
@@ -0,0 +1,502 @@
+        self.misses += 1
+        size = await self._store.getsize(key)
+
+        # Try to get and cache the value if it's reasonably small i.e. ≤10% of max cache size
```
What was the motivation for you choosing this threshold for the value size?

One of the ideas of the cache store is it's useful for stores (e.g., remote stores) where making a query for the store is the reason for them being slow, not then getting a value from the store. So my thinking was it `getsize(key)` takes just as long as `getvalue(key)` for a remote store, the implementation of `LRUCacheStore.getsize` might as well 1) get the value 2) cache it, and 3) 'manually' calculate and return the size of the value, instead of relying on the underlying `_store.getsize(key)` implementation.

Please could you do some benchmarking using a remote store to see if it makes sense to do what I suggested above?

ruaridhg replied:
I changed this now and did some quick benchmarking so your suggestion is correct that the time saved is worth the new implementation.

dstansby commented on tests/test_store/test_cache.py:
```diff
@@ -0,0 +1,471 @@
+        # docstring inherited
+        self._store_dict.clear()
+
+    def __getitem__(self, key: Any) -> Any:
```
Why have you included, and are testing lots of methods (including this one) that are not methods that are defined on the Store base class?

dstansby commented on tests/test_store/test_cache.py:
```diff
@@ -0,0 +1,471 @@
+        return {"store": MemoryStore(), "max_size": 2**27}
+
+    def create_store(self, **kwargs: Any) -> LRUStoreCache:
+        # wrapper therefore no dimension_separator argument
```
I don't understand this comment?

ruaridhg replied:
Removed.

dstansby commented on tests/test_store/test_cache.py:
```diff
@@ -0,0 +1,471 @@
+            raise
+
+
+def skip_if_nested_chunks(**kwargs: Any) -> None:
```
Why do some tests not work with a dimension separator of "/"? As far as I know they definitely should!

ruaridhg replied:
Think this was a hangover from the dict-like implementation which has now been removed.

ruaridhg commented:
> I have some more questions, left inline. As well as checking the tests run, can you check that `pre-commit` runs too, and fix any errors? It looks like there's currently a few typing errors that need addressing in the new code.

I've run pre-commit for the 2 files I changed i.e. _cache.py and the tests

dstansby commented on caching_store.py:
Looks like you accidentally committed this file?

dstansby commented on uv.lock:
Looks like you accidentally committed this file too?
